src/xrpld/app/tx/detail/MPTokenIssuanceSet.cpp

```diff
@@ -57,13 +81,60 @@ MPTokenIssuanceSet::preflight(PreflightContext const& ctx)
     if (holderID && accountID == holderID)
         return temMALFORMED;
 
-    if (ctx.rules.enabled(featureSingleAssetVault))
+    if (ctx.rules.enabled(featureSingleAssetVault) ||
+        ctx.rules.enabled(featureDynamicMPT))
     {
         // Is this transaction actually changing anything ?
-        if (txFlags == 0 && !ctx.tx.isFieldPresent(sfDomainID))
+        if (txFlags == 0 && !ctx.tx.isFieldPresent(sfDomainID) && !isMutate)
             return temMALFORMED;
     }
 
+    if (ctx.rules.enabled(featureDynamicMPT))
+    {
+        // Holder field is not allowed when mutating MPTokenIssuance
+        if (isMutate && holderID)
+            return temMALFORMED;
+
+        // Can not set flags when mutating MPTokenIssuance
+        if (isMutate && txFlags != 0)
+            return temMALFORMED;
+
+        if (auto const fee = ctx.tx[~sfTransferFee])
+        {
+            if (fee > maxTransferFee)
+                return temBAD_TRANSFER_FEE;
+        }
+
+        if (auto const metadata = ctx.tx[~sfMPTokenMetadata])
+        {
+            if (metadata->length() > maxMPTokenMetadataLength)
+                return temMALFORMED;
+        }
+
+        if (auto const mutableFlags = ctx.tx[~sfMutableFlags])
+        {
+            if (!*mutableFlags ||
+                (*mutableFlags & tfMPTokenIssuanceSetMutableMask))
+                return temINVALID_FLAG;
+
+            // Can not set and clear the same flag
+            if (std::any_of(
+                    mptMutabilityFlags.begin(),
+                    mptMutabilityFlags.end(),
+                    [mutableFlags](auto const& f) {
+                        return (*mutableFlags & f.setFlag) &&
+                            (*mutableFlags & f.clearFlag);
+                    }))
+                return temINVALID_FLAG;
+
+            // Trying to set a non-zero TransferFee and clear MPTCanTransfer
+            // in the same transaction is not allowed.
+            if (auto const fee = ctx.tx[~sfTransferFee];
+                fee && *fee > 0 && (*mutableFlags & tfMPTClearCanTransfer))
+                return temMALFORMED;
```

Comment:
nit: can be simplified
```suggestion
            if (auto const fee = ctx.tx[~sfTransferFee].value_or(0);
                fee && (*mutableFlags & tfMPTClearCanTransfer))
                return temMALFORMED;
```

Reply:
Although the fee is guaranteed to be non-negative when the code reaches here, so that they are functionally equivalent, it is still beneficial to keep the explicit asterisk, which would not need pre-assume the fee to be non-negative.